### PR TITLE
changefeedccl: add randomized roundrip tests for json encoding

### LINKED ISSUE #100828: cdc: roundtrip tests for json encoder
Using https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/changefeedccl/avro_test.go as a model, let's get the JSON encoder to parity with tests that iterate over every type, generate random data, encode it the way a changefeed encodes them, and then assert that the resulting JSON can be cast to the same value.

Jira issue: CRDB-26646

### PR DESCRIPTION
**Pull request by jayshrivastava**
This changes adds a test which encodes CDC events using the JSON encoder and decodes them, asserting that the original datums match the decoded ones.

Release note: None
Epic: None
Closes: https://github.com/cockroachdb/cockroach/issues/100828

### CODE REVIEW

File: pkg/ccl/changefeedccl/encoder_test.go [hunk 1]
```diff
@@ -1112,3 +1116,114 @@ func TestParquetEncoder(t *testing.T) {
...
+	}
+
+	// Add arrays of all the scalar types which are supported.
+	arrayTypesToTest := make([]*types.T, 0, 256)
+	for oid := range types.ArrayOids {
+		arrayTyp := types.OidToType[oid]
+		for _, typ := range typesToTest {
+			switch typ {
+			case types.Jsonb:
+				// Unsupported by sql/catalog/colinfo
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/encoder_test.go:1155]: really?

File: pkg/ccl/changefeedccl/encoder_test.go [hunk 2]
```diff
@@ -1112,3 +1116,114 @@ func TestParquetEncoder(t *testing.T) {
...
+	// Add enums.
+	testEnum := createEnum(
+		tree.EnumValueList{`open`, `closed`},
+		tree.MakeUnqualifiedTypeName(`switch`),
+	)
+	typesToTest = append(typesToTest, testEnum)
+
+	// Generate a test for each type with a random datum of that type.
+	for _, typ := range typesToTest {
+		datum := randgen.RandDatum(rng, typ, false /* nullOk */)
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/encoder_test.go:1176]: should nulls be allowed?

File: pkg/ccl/changefeedccl/encoder_test.go [hunk 3]
```diff
@@ -1112,3 +1116,114 @@ func TestParquetEncoder(t *testing.T) {
...
+
+		// name can be "char" (with quotes), so needs to be escaped.
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/encoder_test.go:1178]: does %q as the format work?

File: pkg/ccl/changefeedccl/encoder_test.go [hunk 4]
```diff
@@ -1112,3 +1116,114 @@ func TestParquetEncoder(t *testing.T) {
...
+		tests = append(tests, randTypeTest)
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			tableDesc, err := parseTableDesc(
+				fmt.Sprintf(`CREATE TABLE "%s" %s`, test.name, test.schema))
+			require.NoError(t, err)
+
+			dab := rowenc.EncDatumRow{rowenc.EncDatum{Datum: tree.NewDInt(1)}, rowenc.EncDatum{Datum: test.datum}}
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/encoder_test.go:1195]: dab? what does that stand for?

**Review by miretskiy — APPROVED** [review]

File: pkg/ccl/changefeedccl/encoder_test.go (hunk 1 above)
Comment (jayshrivastava) [reply pkg/ccl/changefeedccl/encoder_test.go:1155]: Yes

File: pkg/ccl/changefeedccl/encoder_test.go (hunk 3 above)
Comment (jayshrivastava) [reply pkg/ccl/changefeedccl/encoder_test.go:1178]: Unfortunately, it does not.

File: pkg/ccl/changefeedccl/encoder_test.go (hunk 4 above)
Comment (jayshrivastava) [reply pkg/ccl/changefeedccl/encoder_test.go]: I changed it!

File: pkg/ccl/changefeedccl/encoder_test.go (hunk 2 above)
Comment (jayshrivastava) [reply pkg/ccl/changefeedccl/encoder_test.go]: Sure.

**Comment by jayshrivastava** [comment]
bors r=miretskiy

**Pull request merged**